libcnb-test/src/container_context.rs

```diff
@@ -129,6 +222,71 @@ pub struct ContainerContext<'a> {
 }
 
 impl<'a> ContainerContext<'a> {
+    /// Gets the container's log output until the current point in time.
+    ///
+    /// Note: This method will only return logs until the current point in time. It will not
+    /// block until the container stops. Since the output of this method depends on timing, directly
+    /// asserting on it's contents might result in flaky tests.
+    ///
+    /// See: [`logs_follow`](Self::logs_follow) for a blocking alternative.
+    ///
+    /// # Panics
+    /// - When the log output could not be consumed/read.
+    #[must_use]
+    pub fn logs(&self) -> LogOutput {
+        // Bollard forces us to cast to i64
+        #[allow(clippy::cast_possible_wrap)]
+        self.logs_internal(bollard::container::LogsOptions {
+            stdout: true,
+            stderr: true,
+            since: 0,
+            until: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("System time is before UNIX epoch")
+                .as_secs() as i64,
+            tail: "all",
+            ..bollard::container::LogsOptions::default()
+        })
+    }
+
+    /// Gets the container's log output until the container stops.
+    ///
+    /// Note: This method will block until the container stops. If the container never stops by
+    /// itself, your test will hang indefinitely. This is common when the container hosts an HTTP
+    /// service.
+    ///
+    /// See: [`logs`](Self::logs) for a non-blocking alternative.
+    ///
+    /// # Panics
+    /// - When the log output could not be consumed/read.
+    #[must_use]
+    pub fn logs_follow(&self) -> LogOutput {
```

Comment:
I'm not 100% sure about the naming of this method, but I don't know if I can think of something better either.

The name suggests that one can actually follow the logs, which isn't the case, and I think might not be intuitive when reading tests calling it.

If I'm understanding correctly this is more like a `logs_blocking()`, or to focus on the desired outcome of calling the function (rather than the side-effect behaviour), `logs_until exit()`? Though both of those names aren't perfect either.

Any other ideas? :-)

Reply:
I went with `follow` as this is the Docker terminology on both the CLI and API. The fact that it blocks and doesn't allow real-time streaming of the logs is IMO well conveyed through the types (since this doesn't return a `Stream`).

I would leave it as-is unless you feel strongly the name should be changed.

Reply:
@edmorley do you want to propose a name?  As you said,`logs_blocking` and `logs_until_exit` aren't amazing either.

Reply:
I'm also curious - in what scenarios will `.logs()` not include all logs, for containers that do exit? From reading `start_internal()` it seems we block on the `create_container` and `start_container` steps, so presumably by the time the test-provided function runs, the container will have at least started? But I presume the issue is that it's not guaranteed that the started command will have finished running and outputting to stdout/stderr by the time the test block runs?

It just seems like `.logs()` is perhaps something that should rarely be used (unless querying the logs of a web server, and even then careful use of sleeps etc will be needed), and instead `.logs_follow()` should be preferred in most cases?

If so, then I worry that having the names as `.logs()` and `.logs_follow()` doesn't make it clear that the former is a often a footgun and the latter is often the best practice, since:
- To me, when there are several methods with similar names, some with suffixes and some without, then the ones without the suffixes seem like the standard/usual case.
- "Follow" makes me think it's some interactive thing, therefore not what I need, therefore the "other" method is probably what I need.

Now obviously reading the rustdocs does make things clearer, however when people are say reviewing a PR via the GitHub code interface, I think it would be very easy to forget that a `.logs()` usage in a test is likely a footgun.

What about something like:
- `logs_follow()` -> `logs_wait()`
- `logs()` -> `logs_now()`

Or even this, if we see "follow" as being the safer (as in, less prone to races)/more common usage:
- `logs_follow()` -> `logs()`
- `logs()` -> `logs_now()`

(I'm not overly attached to the exact names, more-so the idea of not having the footgun method be the one without the suffix)

Reply:
> Or even this, if we see "follow" as being the more common usage

Thing is, `logs_follow` isn't without it's own foot-guns either. If you call it on a container reference that runs a never-terminating process such as a web service (which isn't uncommon), that call will hang indefinitely without any indication what is going on.

With `logs`, the assert will fail when the test runs into timing issues, but that will at least directly point to the assert as the source of the issue.

I'd say which of the two you want to use is really dependent on the use-case: if it's a long running process, you probably want `logs` with some extra timing logic. If we're talking about a one-off command,`logs_follow` is easier to use since the test doesn't need any custom timing logic. There is no "default".

Honestly, I still don't think the naming of the log functions is bad and/or misleading. The naming and behaviour exactly mirrors the Docker CLI when getting logs from a running container and the Docker API (which is used programatically, not as an interactive process) uses the same naming too.

Even if there is a slightly better name, we always pay the price of using a different name than the underlying (and not abstracted away) technology. I feel like we already spent way too much time on this minor naming discussion which will probably have no measurable impact - I'm sure people will run into the timing/blocking issues with either name when they don't read the docs that mention the caveats.

Reply:
That said, to cut this short I'm willing to rename them as proposed if that ends this bikeshedding thread:

```
logs_follow() -> logs_wait()
logs() -> logs_now()
```

Reply:
> If you call it on a container reference that runs a never-terminating process such as a web service (which isn't uncommon), that call will hang indefinitely without any indication what is going on.

I agree, but this issue is (a) deterministic (will hang every time), (b) will be visible whilst writing/developing the test, not at some random point later.

> I'd say which of the two you want to use is really dependent on the use-case: if it's a long running process, you probably want `logs` with some extra timing logic.

I absolutely agree, I was just wondering if this case was less common.

> Honestly, I still don't think the naming of the log functions is bad and/or misleading. The naming and behaviour exactly mirrors the Docker CLI when getting logs from a running container and the Docker API (which is used programatically, not as an interactive process) uses the same naming too.

So whilst sometimes looking at the underlying terminology can be useful (ie: we don't want to make concepts up when adding layers of abstraction, otherwise it can add to the mental overhead to understand the overall model), I think in this case, the users of libcnb-test aren't thinking in terms of the Docker CLI or API, but a test runner letting them make assertions in a testcase. I think that changes things slightly.

> That said, to cut this short I'm willing to rename them as proposed if that ends this bikeshedding thread:

Thank you for making the change. However I think it's a little unfair to call a genuine concern about users mis-using an API due to naming (which can result in either hangs or flaky tests; this isn't just a vanity naming thing), merely "bikeshedding" :-)